### ISSUE TITLE
Compile error after panic can cause "Variable 'x' has gone missing"

### ISSUE THREAD

**Issue by j-tai** [google/evcxr#92]
If you try to access a variable immediately after a panic, then type `:vars`, the variable will come back from the dead. Then, trying to access it again will result in a panic "Variable 'x' has gone missing".

```
>> let x = 42;
>> panic!()
thread '<unnamed>' panicked at 'explicit panic', src/lib.rs:105:1
[...]
Child process terminated with status: exit code: 101
>> x
   ^ not found in this scope
cannot find value `x` in this scope
>> :vars
x: i32

>> x
thread '<unnamed>' panicked at 'Variable 'x' has gone missing', src/lib.rs:64:21
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1028
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:188
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:205
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:464
  11: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:373
  12: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:328
  13: ctx::evcxr_internal_runtime::VariableStore::take_variable
  14: run_user_code_4
  15: evcxr::runtime::Runtime::run_loop
  16: evcxr::runtime::runtime_hook
  17: evcxr::main
  18: std::rt::lang_start::{{closure}}
  19: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:48
  20: std::panicking::try::do_call
             at src/libstd/panicking.rs:287
  21: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  22: std::panicking::try
             at src/libstd/panicking.rs:265
  23: std::panic::catch_unwind
             at src/libstd/panic.rs:396
  24: std::rt::lang_start_internal
             at src/libstd/rt.rs:47
  25: main
  26: __libc_start_main
  27: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Child process terminated with status: exit code: 101
```

evcxr version: 0.4.7

**Comment by bjorn3** [comment]
This is expected. You can run `:preserve_vars_on_panic 1` to keep them after a panic. This will be slower though.

**Comment by davidlattimore** [comment]
Thanks for reporting that. The loss of the variable on panic is expected, but the subsequent panic on trying to access the variable isn't expected. The variable is supposed to be removed from the list of variables, so subsequent access attempts should result in a compilation error, not another panic.

**Comment by davidlattimore** [comment]
Hmm, I'm having trouble reproducing. There must be something I'm missing, because I do get the compilation error that I'd have expected.

```
>> x
   ^ not found in this scope
cannot find value `x` in this scope
```

**Comment by j-tai** [comment]
It turns out you don't need to try to access the variable; any compile error will do.

I'm running this on GNU/Linux, if that helps.

Here's a list of inputs to reproduce it:

* `let x = 42;` (create a variable)
* `panic!()` (panic)
* `?` (any compile error)
* `:vars` (list the variables; this is a necessary step)
* `x` (access the variable again)

If you're still having trouble reproducing, try running evcxr with this Dockerfile:

```dockerfile
FROM rust:1.40
RUN cargo install evcxr_repl
CMD ["evcxr_repl"]
```

(Run it with `docker run -it image_id`)

**Comment by davidlattimore** [comment]
Ah, thanks. Yep, I can reproduce now. Pretty sure I know what's happening. The state (which includes which variables are defined) is copied when execution is successful, then when a compilation error is encountered, it rolls back to the last committed state. When a panic occurs, the variables are removed, but that change must not get committed.

I think I used to copy the current state at the start of execution then roll back to that copy on compilation failure, but for some reason I changed it to store the committed state. I should probably try to work out why I made that change to avoid regressing it. Hopefully I added a test. In any case, this probably shouldn't be too hard to fix. Thanks again for the report.
